src/cid.js

```diff
@@ -0,0 +1,284 @@
+   *      given.
+   */
+  get(externalCidScope, consent, opt_persistenceConsent) {
+    return consent.then(() => {
+      return getExternalCid(this, externalCidScope,
+          opt_persistenceConsent || consent);
+    });
+  }
+}
+
+/**
+ * Returns the "external cid". This is a cid for a specific purpose
+ * (Say Analytics provider X). It is unique per user, that purpose
+ * and the AMP origin site.
+ * @param {!Cid} cid
+ * @param {string} externalCidScope
+ * @param {!Promise} persistenceConsent
+ * @return {!Promise<string|null>}
+ */
+function getExternalCid(cid, externalCidScope, persistenceConsent) {
+  const url = parseUrl(cid.win.location.href);
+  if (!isProxyOrigin(url)) {
+    return new Promise(resolve => {
+      resolve(getCookie(cid.win, externalCidScope));
+    });
+  }
+  return getBaseCid(cid, persistenceConsent).then(baseCid => {
+    return cid.sha384Base64_(
+        baseCid +
+        getSourceOrigin(url) +
+        externalCidScope);
+  });
+}
+
+/**
+ * Returns whether the URL has the origin of a proxy
+ * @param {Url} url URL of an AMP document.
+ * @return {boolean}
+ * @visibleForTesting BUT if this is needed elsewhere it could be
+ *     factored into its own package.
+ */
+export function isProxyOrigin(url) {
+  return (url.origin == 'https://cdn.ampproject.org' ||
+      url.origin.indexOf('http://localhost:') == 0);
+}
+
+/**
+ * Returns the source origin of an AMP document for documents served
+ * on a proxy origin. Throws if the doc is not on a proxy origin.
+ * @param {Url} url URL of an AMP document.
+ * @return {string} The source origin of the URL.
+ * @visibleForTesting BUT if this is needed elsewhere it could be
+ *     factored into its own package.
+ */
+export function getSourceOrigin(url) {
+  // List of well known proxy hosts. New proxies must be added here
+  // to generate correct tokens.
+  assert(isProxyOrigin(url), 'Expected proxy origin %s', url.origin);
+  // Example path that is being matched here.
+  // https://cdn.ampproject.org/c/s/www.origin.com/foo/
+  // The /s/ is optional and signals a secure origin.
+  const path = url.pathname.split('/');
+  const prefix = path[1];
+  assert(prefix == 'c' || prefix == 'v',
+      'Unknown path prefix in url %s', url.href);
+  const domainOrHttpsSignal = path[2];
+  const origin = domainOrHttpsSignal == 's'
+      ? 'https://' + path[3]
+      : 'http://' + domainOrHttpsSignal;
+  // Sanity test that what we found looks like a domain.
+  assert(origin.indexOf('.') > 0, 'Expected a . in origin %s Foo', origin);
+  return origin;
+}
+
+/**
+ * Returns the base cid for the current user. This string must not
+ * be exposed to users without hashing with the current source origin
+ * and the externalCidScope.
+ * On a proxy this value is the same for a user across all source
+ * origins.
+ * @param {!Cid} cid
+ * @param {!Promise} persistenceConsent
+ * @return {!Promise}
+ */
+function getBaseCid(cid, persistenceConsent) {
+  if (cid.baseCid_) {
```

Comment:
Right. If you keep your browser tap open for a year I think that is fine to count it as a single session.